test/e2e/storage/testsuites/provisioning.go

```diff
@@ -283,11 +313,54 @@ func TestDynamicProvisioning(t StorageClassTest, client clientset.Interface, cla
 			command += fmt.Sprintf(" && ( mount | grep 'on /mnt/test' | awk '{print $6}' | sed 's/^(/,/; s/)$/,/' | grep -q ,%s, )", option)
 		}
 		command += " || (mount | grep 'on /mnt/test'; false)"
-		runInPodWithVolume(client, claim.Namespace, claim.Name, t.NodeName, command, t.NodeSelector, t.ExpectUnschedulable)
+
+		// This test uses TestConfig.ClientNodeName and TestConfig.ClientNodeSelector
+		// as follows:
+		// - first it runs two pods on the same node, selected based on
+		//   TestConfig.ClientNodeName (if not empty) or TestConfig.ClientNodeSelector
+		//   (otherwise)
+		// - if TestConfig.ClientNodeSelector matches more than the node from
+		//   the first step, it will run a third pod on a different node
+
+		By("checking the created volume is writable")
+		runInPodWithVolume(client, claim.Namespace, claim.Name, "-first", nodeName, command,
+			nil /* node selector */, false /* expect unschedulable */)
 
 		By("checking the created volume is readable and retains data")
-		runInPodWithVolume(client, claim.Namespace, claim.Name, t.NodeName, "grep 'hello world' /mnt/test/data", t.NodeSelector, t.ExpectUnschedulable)
+		runInPodWithVolume(client, claim.Namespace, claim.Name, "-second", nodeName, "grep 'hello world' /mnt/test/data", nil /* node selector */, t.ExpectUnschedulable)
+
+		// Run on another node, if we have one.
+		var secondNodeName string
+		if nodes != nil {
+			for _, node := range nodes.Items {
+				if node.Name != nodeName {
+					secondNodeName = node.Name
+					break
+				}
+			}
+		}
+		if secondNodeName != "" {
+			By("checking the created volume is readable on another node")
+			runInPodWithVolume(client, claim.Namespace, claim.Name, "-third", secondNodeName, "grep 'hello world' /mnt/test/data", nil /* node selector */, t.ExpectUnschedulable)
+		}
 	}
+
+	if t.MultiWriteCheck {
+		// We start two pods concurrently on the same node,
+		// using the same PVC. Both wait for other to create a
+		// file before returning.
+		wg := sync.WaitGroup{}
+		wg.Add(2)
+		run := func(suffix, command string) {
+			defer GinkgoRecover()
+			defer wg.Done()
+			runInPodWithVolume(client, claim.Namespace, claim.Name, suffix, nodeName, command, nil /* node selector */, t.ExpectUnschedulable)
+		}
+		go run("-first", "touch /mnt/test/first && while ! [ -f /mnt/test/second ]; do sleep 1; done")
+		go run("-second", "touch /mnt/test/second && while ! [ -f /mnt/test/first ]; do sleep 1; done")
+		wg.Wait()
```

Comment:
should we have a timeout here? how long is the default timeout for a long test (if there is one)?

Reply:
`runInPodWithVolume` has its own timeout, 5 minutes I think (from ` framework.WaitForPodSuccessInNamespaceSlow`) - basically the same as in all the other tests. I don't think we need another one.